File: iot/amqp/header/api/AMQPWrapper.py
import iot.amqp.avps.AMQPType as AMQPType 
import iot.amqp.tlv.impl.TLVNull as TLVNull
import iot.amqp.tlv.impl.TLVArray as TLVArray 
import iot.amqp.tlv.impl.TLVList as TLVList
import iot.amqp.tlv.impl.TLVMap as TLVMap
import iot.amqp.tlv.impl.TLVFixed as TLVFixed
import iot.amqp.tlv.impl.TLVVariable as TLVVariable
import iot.amqp.numeric.NumericUtil as NumericUtil

class amqpWrapper():
    def __init__(self):
        self.amqpType = AMQPType.amqpType()

    def wrap(self, obj):
        result = None
        if obj is None:
            return TLVNull.tlvNull()

        if obj.getType() == 'ubyte':
            result = self.wrapByte(obj.getValue())
        elif obj.getType() == 'ushort':
            if obj.getValue() >= 0:
                result = self.wrapUByte(obj.getValue())
            else:
                result = self.wrapShort(obj.getValue())
        elif obj.getType() == 'uint':
            if obj.getValue() > 0:
                result = self.wrapUShort(obj.getValue())
            else:
                result = self.wrapInt(obj.getValue())
        elif obj.getType() == 'ulong':
            if obj.getValue() >= 0:
                result = self.wrapUInt(obj.getValue())
            else:
                result = self.wrapLong(obj.getValue())
        elif obj.getType() == 'decimal128':
            result = self.wrapULong(obj.getValue())
        elif obj.getType() == 'string':
            result = self.wrapString(obj.getValue())
        elif obj.getType() == 'AMQPSymbol':
            result = self.wrapSymbol(obj.getValue())
        elif obj.getType() == 'bytearray':
            result = self.wrapBinary(obj.getValue())
        elif obj.getType() == 'bool':
            result = self.wrapBool(obj.getValue())
        elif obj.getType() == 'chararray':
            result = self.wrapChar(obj.getValue())
        elif obj.getType() == 'double':
            result = self.wrapDouble(obj.getValue())
        elif obj.getType() == 'float':
            result = self.wrapFloat(obj.getValue())
        elif obj.getType() == 'uuid':
            result = self.wrapUuid(obj.getValue())
        elif obj.getType() == 'timestamp':
            result = self.wrapTimestamp(obj.getValue())
        elif obj.getType() == 'AMQPDecimal':
            val = obj.getValue()
            if len(val) == 4:
                result = self.wrapDecimal32(val)
            elif len(val) == 8:
                result = self.wrapDecimal64(val)
            elif len(val) == 16:
                result = self.wrapDecimal128(val)
        else:
            print('Wrapper received unrecognized type ' + str(obj))

        return result

    def wrapBool(self, bool):
        value = bytearray()
        if bool:
            code = self.amqpType.getValueByKey('BOOLEAN_TRUE')
        else:
            code = self.amqpType.getValueByKey('BOOLEAN_FALSE')
        return TLVFixed.tlvFixed(code, value)

    def wrapUByte(self, bt):
        if bt < 0:
            print('Error negative value of ' + str(bt) + ' cannot be assigned to UBYTE type')
        return TLVFixed.tlvFixed(self.amqpType.getValueByKey('UBYTE'), bt)

    def wrapByte(self, bt):
        return TLVFixed.tlvFixed(self.amqpType.getValueByKey('BYTE'), bt)

    def wrapUInt(self, i):
        if i < 0:
            print('Error negative value of ' + str(i) + ' cannot be assigned to UINT type')
        value = self.convertUInt(i)
        code = None
        if len(value) == 0:
            code = self.amqpType.getValueByKey('UINT_0')
        elif len(value) == 1:
            code = self.amqpType.getValueByKey('SMALL_UINT')
        elif len(value) > 1:
            code = self.amqpType.getValueByKey('UINT')
        return TLVFixed.tlvFixed(code, value)

    def wrapInt(self, i):
        value = self.convertUInt(i)
        code = None
        if len(value) > 0:
            code = self.amqpType.getValueByKey('INT')
        else:
            code = self.amqpType.getValueByKey('SMALL_INT')
        return TLVFixed.tlvFixed(code, value)

    def wrapULong(self, l):
        if l is None:
            print('Wrapper cannot wrap ulong null')
        if l < 0:
            print('negative value of ' + str(l) + ' cannot be assignet to ULONG type')
        value = self.convertULong(l)
        code = None
        if len(value) == 0:
            code = self.amqpType.getValueByKey('ULONG_0')
        elif len(value) == 1:
            code = self.amqpType.getValueByKey('SMALL_ULONG')
        else:
            code = self.amqpType.getValueByKey('ULONG')
        return TLVFixed.tlvFixed(code, value)

    def wrapLong(self, l):
        value = self.convertULong(l)
        code = None
        if len(value) > 1:
            code = self.amqpType.getValueByKey('LONG')
        else:
            code = self.amqpType.getValueByKey('SMALL_LONG')
        return TLVFixed.tlvFixed(code, value)

    def wrapBinary(self, bt):
        if bt is None:
            print('Wrapper cannot wrap binary null')
        code = None
        if len(bt) > 255:
            code = self.amqpType.getValueByKey('BINARY_32')
        else:
            code = self.amqpType.getValueByKey('BINARY_8')
        return TLVFixed.tlvFixed(code, bt)

    def wrapUuid(self, uuid):
        if uuid is None:
            print('Wrapper cannot wrap uuid null')
        return TLVFixed.tlvFixed(self.amqpType.getValueByKey('UUID'), bytes(uuid, 'utf-8'))

    def wrapUShort(self, sh):
        if sh < 0:
            print('negative value of ' + str(sh) + ' cannot be assignet to UShort type')
        data = bytearray()
        data = NumericUtil.util.addShort(data, sh)
        return TLVFixed.tlvFixed(self.amqpType.getValueByKey('USHORT'), data)

    def wrapShort(self, sh):
        data = bytearray()
        data = NumericUtil.util.addShort(data, sh)
        return TLVFixed.tlvFixed(self.amqpType.getValueByKey('USHORT'), data)

    def wrapDouble(self, db):
        data = bytearray()
        data = NumericUtil.util.addDouble(data, db)
        return TLVFixed.tlvFixed(self.amqpType.getValueByKey('DOUBLE'), data)

    def wrapFloat(self, f):
        data = bytearray()
        data = NumericUtil.util.addFloat(data, f)
        return TLVFixed.tlvFixed(self.amqpType.getValueByKey('FLOAT'), data)

    def wrapChar(self, ch):
        data = bytearray()
        data = NumericUtil.util.addInt(data, ch)
        return TLVFixed.tlvFixed(self.amqpType.getValueByKey('CHAR'), data)

    def wrapTimestamp(self, stamp):
        if stamp is None:
            print('Wrapper cannot wrap timestamp null')
        data = bytearray()
        data = NumericUtil.util.addLong(data, stamp)
        return TLVFixed.tlvFixed(self.amqpType.getValueByKey('TIMESTAMP'), data)

    def wrapDecimal32(self, d):
        if d is None:
            print('Wrapper cannot wrap decimal32 null')
        return TLVFixed.tlvFixed(self.amqpType.getValueByKey('DECIMAL_32'), d.getValue())

    def wrapDecimal64(self, d):
        if d is None:
            print('Wrapper cannot wrap decimal64 null')
        return TLVFixed.tlvFixed(self.amqpType.getValueByKey('DECIMAL_64'), d.getValue())

    def wrapDecimal128(self, d):
        if d is None:
            print('Wrapper cannot wrap decimal128 null')
        return TLVFixed.tlvFixed(self.amqpType.getValueByKey('DECIMAL_128'), d.getValue())

    def wrapString(self, s):
        if s is None:
            print('Wrapper cannot wrap string null')
        value = bytes(s,'utf-8')
        code = None
        if len(value) > 255:
            code = self.amqpType.getValueByKey('STRING_32')
        else:
            code = self.amqpType.getValueByKey('STRING_8')
        return TLVVariable.tlvVariable(code, value)

    def wrapSymbol(self, s):
        if s is None:
            print('Wrapper cannot wrap symbol null')
        value = bytes(s.getValue(),'utf-8')
        code = None
        if len(value) > 255:
            code = self.amqpType.getValueByKey('SYMBOL_32')
        else:
            code = self.amqpType.getValueByKey('SYMBOL_8')
        return TLVVariable.tlvVariable(code, value)

    def wrapList(self, input):
        if input is None:
            print('Wrapper cannot wrap null list')
        list = TLVList.tlvList(None,None)
        for obj in input:
            list.addElement(self.wrap(obj))
        return list

    def wrapMap(self, input):
        if input is None:
            print('Wrapper cannot wrap null map')
        map = TLVMap.tlvMap(None,None)
        if isinstance(input, dict):
            for key, value in input.items():
                map[self.wrap(key)] = self.wrap(value)
        return map

    def wrapArray(self, input):
        if input is None:
            print('Wrapper cannot wrap null array')
        array = TLVArray.tlvArray(None,None)
        for obj in input:
            res = self.wrap(obj)
            array.addElement(res)
        return array

    def convertUInt(self, i):
        data = bytearray()
        if i == 0:
            return data
        elif i > 0 and i <= 255:
            data = NumericUtil.util.addByte(data, i)
            return data
        else:
            data = NumericUtil.util.addInt(data, i)
            return data

    def convertInt(self, i):
        data = bytearray()
        if i == 0:
            return data
        elif i >= -128 and i <= 127:
            return i
        else:
            data = NumericUtil.util.addInt(data, i)
            return data

    def convertULong(self, l):
        data = bytearray()
        if l == 0:
            return data
        elif l > 0 and l <= 255:
            return l
        else:
            data = NumericUtil.util.addLong(data, l)
            return data

    def convertLong(self, l):
        data = bytearray()
        if l == 0:
            return data
        elif l >= -128 and l <= 127:
            return l
        else:
            data = NumericUtil.util.addLong(data, l)
            return data

File: IoTbrokerClient/src/iot-protocols/amqp/classes/headerapi/amqpwrapper.cpp
/**
 * Mobius Software LTD
 * Copyright 2015-2018, Mobius Software LTD
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

#include "amqpwrapper.h"

TLVAMQP *AMQPWrapper::wrapObject(JsonVariant *object)
{
    if (object == NULL) {
        return new AMQPTLVNull();
    }

    TLVAMQP *result = NULL;

    AMQPSimpleType *type = (AMQPSimpleType *)object->value<AMQPSimpleType *>();

    switch (type->getType()) {
        case AMQP_BOOLEAN_TYPE:
        {
            result = AMQPWrapper::wrapBool(AMQPSimpleType::variantToBool(std::get<byte>(object));
        }  break;
        case AMQP_UBYTE_TYPE:
        {
            result = AMQPWrapper::wrapUByte(AMQPSimpleType::variantToUByte(std::get<unsigned char>(object)));
        }  break;
        case AMQP_USHORT_TYPE:
        case AMQP_UINT_TYPE:
        case AMQP_SMALL_UINT_TYPE:
        case AMQP_UINT0_TYPE:
        {
            result = AMQPWrapper::wrapUShort(AMQPSimpleType::variantToUShort(std::get<unsigned short>(object)));
        }  break;
        case AMQP_ULONG_TYPE:
        case AMQP_SMALL_ULONG_TYPE:
        case AMQP_ULONG0_TYPE:
        {
            result = AMQPWrapper::wrapUInt(AMQPSimpleType::variantToUInt(std::get<unsigned int>(object)));
        }  break;
        case AMQP_BYTE_TYPE:
        {
            result = AMQPWrapper::wrapByte(AMQPSimpleType::variantToChar(std::get<unsigned char>(object)));
        }  break;
        case AMQP_SHORT_TYPE:
        {
            result = AMQPWrapper::wrapShort(AMQPSimpleType::variantToShort(std::get<short>(object)));
        }  break;
        case AMQP_INT_TYPE:
        case AMQP_SMALL_INT_TYPE:
        {
            result = AMQPWrapper::wrapInt(AMQPSimpleType::variantToInt(std::get<int>(object)));
        }  break;
        case AMQP_LONG_TYPE:
        case AMQP_SMALL_LONG_TYPE:
        {
            result = AMQPWrapper::wrapLong(AMQPSimpleType::variantToLong(std::get<long>(object)));
        }  break;
        case AMQP_FLOAT_TYPE:
        {
            result = AMQPWrapper::wrapFloat(AMQPSimpleType::variantToFloat(std::get<float>(object)));
        }  break;
        case AMQP_DOUBLE_TYPE:
        {
            result = AMQPWrapper::wrapDouble(AMQPSimpleType::variantToDouble(std::get<double>(object)));
        }  break;
        case AMQP_CHAR_TYPE:
        {
            result = AMQPWrapper::wrapChar(AMQPSimpleType::variantToChar(std::get<char>(object)));
        }
        case AMQP_DECIMAL32_TYPE:
        case AMQP_DECIMAL64_TYPE:
        case AMQP_DECIMAL128_TYPE:
        {
            AMQPDecimal *decimal = AMQPSimpleType::variantToDecimal(object);
            if (sizeof(decimal->getValue()) == 4) {
                result = AMQPWrapper::wrapDecimal32(decimal);
            } else if (sizeof(decimal->getValue()) == 8) {
                result = AMQPWrapper::wrapDecimal64(decimal);
            } else if (sizeof(decimal->getValue()) == 16) {
                result = AMQPWrapper::wrapDecimal128(decimal);
            }
        }  break;
        case AMQP_TIMESTAMP_TYPE:
        {
            result = AMQPWrapper::wrapTimestamp(AMQPSimpleType::variantToTimestamp(std::get<time_t>(object)));
        }  break;
        case AMQP_UUID_TYPE:
        {
            result = AMQPWrapper::wrapUUID(AMQPSimpleType::variantToUuid(std::get<String>(object)));
        }  break;
        case AMQP_BINARY8_TYPE:
        case AMQP_BINARY32_TYPE:
        {
            result = AMQPWrapper::wrapBinary(AMQPSimpleType::variantToBinary(std::get<unsigned char *>(object)));
        }  break;
        case AMQP_STRING8_TYPE:
        case AMQP_STRING32_TYPE:
        {
            result = AMQPWrapper::wrapString(AMQPSimpleType::variantToString(std::get<String>(object)));
        }  break;
        case AMQP_SYMBOL8_TYPE:
        case AMQP_SYMBOL32_TYPE:
        {
            result = AMQPWrapper::wrapSymbol(AMQPSimpleType::variantToSymbol(std::get<char>(object)));
        }  break;
        default: printf("UNWRAPPER: wrong type!"; break);
    }

    return result;
}

TLVAMQP *AMQPWrapper::wrapBool(bool value)
{
    unsigned char * data = (unsigned char *) malloc(sizeof(char));
    AMQPType *type = new AMQPType(value ? AMQP_BOOLEAN_TRUE_TYPE : AMQP_BOOLEAN_FALSE_TYPE);
    return new AMQPTLVFixed(type, data);
}

TLVAMQP *AMQPWrapper::wrapUByte(short value)
{
    if (value < 0) {
        printf("AMQPWrapper::wrapUByte");
        return NULL;
    }

    AMQPType *type = new AMQPType(AMQP_UBYTE_TYPE);
    unsigned char * data = (unsigned char *) malloc(sizeof(char));
    data[0] = value;
    return new AMQPTLVFixed(type, data);
}

TLVAMQP *AMQPWrapper::wrapByte(char value)
{
    AMQPType *type = new AMQPType(AMQP_BYTE_TYPE);
    unsigned char * data = (unsigned char *) malloc(sizeof(char));
    data[0] = value;
    return new AMQPTLVFixed(type, data);
}

TLVAMQP *AMQPWrapper::wrapUInt(unsigned int value)
{
    unsigned char * data = AMQPWrapper::convertUInt(value);
    AMQPType *type = NULL;

    if (sizeof(data) == 0) {
        type = new AMQPType(AMQP_UINT0_TYPE);
    } else if (sizeof(data) == 1) {
        type = new AMQPType(AMQP_SMALL_UINT_TYPE);
    } else if (sizeof(data) > 1) {
        type = new AMQPType(AMQP_UINT_TYPE);
    }
    return new AMQPTLVFixed(type, data);
}

TLVAMQP *AMQPWrapper::wrapInt(int value)
{
    unsigned char * data = AMQPWrapper::convertInt(value);
    AMQPTypes tp = (sizeof(data) > 1) ? AMQP_INT_TYPE : AMQP_SMALL_INT_TYPE;
    AMQPType *type = new AMQPType(tp);
    return new AMQPTLVFixed(type, data);
}

TLVAMQP *AMQPWrapper::wrapULong(unsigned long value)
{
    unsigned char * data = AMQPWrapper::convertULong(value);
    AMQPType *type = NULL;

    if (sizeof(data) == 0) {
        type = new AMQPType(AMQP_ULONG0_TYPE);
    } else if (sizeof(data) == 1) {
        type = new AMQPType(AMQP_SMALL_ULONG_TYPE);
    } else if (sizeof(data) > 1) {
        type = new AMQPType(AMQP_ULONG_TYPE);
    }
    return new AMQPTLVFixed(type, data);
}

TLVAMQP *AMQPWrapper::wrapLong(long value)
{
    unsigned char * data = AMQPWrapper::convertLong(value);
    AMQPTypes tp = (sizeof(data) > 1) ? AMQP_LONG_TYPE : AMQP_SMALL_LONG_TYPE;
    AMQPType *type = new AMQPType(tp);
    return new AMQPTLVFixed(type, data);
}

AMQPTLVVariable *AMQPWrapper::wrapBinary(unsigned char * value)
{
    if (value == NULL) {
        printf("AMQPWrapper::wrapBinary";)
        return NULL;
    }

    AMQPTypes tp = (sizeof(value) > 255) ? AMQP_BINARY32_TYPE : AMQP_BINARY8_TYPE;
    AMQPType *type = new AMQPType(tp);
    return new AMQPTLVVariable(type, value);
}

TLVAMQP *AMQPWrapper::wrapUUID(String uuid)
{
    if (uuid == NULL) {
        printf("AMQPWrapper::wrapUUID");
        return NULL;
    }

    AMQPType *type = new AMQPType(AMQP_UUID_TYPE);
    return new AMQPTLVFixed(type, uuid);
}

TLVAMQP *AMQPWrapper::wrapUShort(unsigned short value)
{
    if (value < 0) {
        printf("AMQPWrapper::wrapUShort");
        return NULL;
    }

    AMQPType *type = new AMQPType(AMQP_USHORT_TYPE);
    unsigned char * data = malloc(sizeof(short));
    add_short(&data[0], value);
    return new AMQPTLVFixed(type, data);
}

TLVAMQP *AMQPWrapper::wrapShort(short value)
{
    AMQPType *type = new AMQPType(AMQP_SHORT_TYPE);
    unsigned char * data = malloc(sizeof(short));
    add_short(&data[0], value);
    return new AMQPTLVFixed(type, data);
}

TLVAMQP *AMQPWrapper::wrapDouble(double value)
{
    AMQPType *type = new AMQPType(AMQP_DOUBLE_TYPE);

    unsigned char * data = malloc(sizeof(double));
    add_double(&data[0], value);
    return new AMQPTLVFixed(type, data);
}

TLVAMQP *AMQPWrapper::wrapFloat(float value)
{
    AMQPType *type = new AMQPType(AMQP_FLOAT_TYPE);
    unsigned char * data = malloc(sizeof(float));
    add_float(&data[0], value);
    return new AMQPTLVFixed(type, data);
}

TLVAMQP *AMQPWrapper::wrapChar(char value)
{
    AMQPType *type = new AMQPType(AMQP_CHAR_TYPE);
    unsigned char * data = malloc(sizeof(char));
    data[0] = value;
    return new AMQPTLVFixed(type, data);
}

TLVAMQP *AMQPWrapper::wrapTimestamp(time_t value)
{
    if (value == NULL) {
        printf("AMQPWrapper::wrapTimestamp");
        return NULL;
    }
    AMQPType *type = new AMQPType(AMQP_TIMESTAMP_TYPE);
    std::stringstream ss;
    ss << value;
    return new AMQPTLVFixed(type, ss.str());
}

TLVAMQP *AMQPWrapper::wrapDecimal32(AMQPDecimal *value)
{
    if (value == NULL) {
        printf("AMQPWrapper::wrapDecimal32");
        return NULL;
    }

    AMQPType *type = new AMQPType(AMQP_DECIMAL32_TYPE);
    return new AMQPTLVFixed(type, value->getValue());
}

TLVAMQP *AMQPWrapper::wrapDecimal64(AMQPDecimal *value)
{
    if (value == NULL) {
        printf("AMQPWrapper::wrapDecimal64");
        return NULL;
    }

    AMQPType *type = new AMQPType(AMQP_DECIMAL64_TYPE);
    return new AMQPTLVFixed(type, value->getValue());
}

TLVAMQP *AMQPWrapper::wrapDecimal128(AMQPDecimal *value)
{
    if (value == NULL) {
        printf("AMQPWrapper::wrapDecimal128");
        return NULL;
    }

    AMQPType *type = new AMQPType(AMQP_DECIMAL128_TYPE);
    return new AMQPTLVFixed(type, value->getValue());
}

AMQPTLVVariable *AMQPWrapper::wrapString(String value)
{
    if (value == NULL) {
        printf("AMQPWrapper::wrapString");
        return NULL;
    }

    AMQPTypes tp = (sizeof(value) > 255) ? AMQP_STRING32_TYPE : AMQP_STRING8_TYPE;
    AMQPType *type = new AMQPType(tp);
    return new AMQPTLVVariable(type, value);
}

AMQPTLVVariable *AMQPWrapper::wrapSymbol(AMQPSymbol *value)
{
    if (value == NULL) {
        printf("AMQPWrapper::wrapSymbol");
        return NULL;
    }

    unsigned char * bytes = value->getValue();
    AMQPTypes tp = (sizeof(bytes) > 255) ? AMQP_SYMBOL32_TYPE : AMQP_SYMBOL8_TYPE;
    AMQPType *type = new AMQPType(tp);
    return new AMQPTLVVariable(type, bytes);
}

AMQPTLVList *AMQPWrapper::wrapList(std::list<JsonVariant *> value)
{
    if (value.size() == 0) {
        printf("AMQPWrapper::wrapList");
        return NULL;
    }

    AMQPTLVList *list = new AMQPTLVList();

    std::list<JsonVariant>::iterator it;
    for(it = list.begin(); it != list.end(); ++it) {
        list->addElement(AMQPWrapper::wrapObject(it));
    }
    return list;
}

AMQPTLVMap *AMQPWrapper::wrapMap(std::map<JsonVariant *, JsonVariant *> value)
{
    if (value.count() == 0) {
        printf("AMQPWrapper::wrapMap");
        return NULL;
    }

    AMQPTLVMap *map = new AMQPTLVMap();

    for (std::map<JsonVariant *, JsonVariant *>::iterator it = value.begin(); it != value.end(); ++it) {
        JsonVariant *vl = it.second;
        map->putElement(AMQPWrapper::wrapObject(it.first), AMQPWrapper::wrapObject(vl));
    }
    return map;
}

AMQPTLVArray *AMQPWrapper::wrapArray(std::list<JsonVariant *> value)
{
    if (value.count() == 0) {
        printf("AMQPWrapper::wrapArray");
        return NULL;
    }

    AMQPTLVArray *array = new AMQPTLVArray();
    std::list<JsonVariant>::iterator it;
    for(it = list.begin(); it != list.end(); ++it) {
        array->addElement(AMQPWrapper::wrapObject(it));
    }
    return array;
}

// private methods

unsigned char *AMQPWrapper::convertUInt(unsigned int number)
{
    unsigned char * data = malloc(sizeof(unsigned int));

    if (number == 0) {
        return data;
    } else if (number > 0 && number <= 255) {
        data[0] = number;
        return data;
    } else {
        add_int(&data[0], number);
        return data;
    }
}

unsigned char *AMQPWrapper::convertInt(int number)
{
    unsigned char * data = malloc(sizeof(int));

    if (number == 0) {
        return data;
    } else if (number >= -128 && number <= 127) {
        data[0] = number;
        return data;
    } else {
        add_int(&data[0], number);
        return data;
    }
}

unsigned char * AMQPWrapper::convertULong(unsigned long number)
{
    unsigned char * data = malloc(sizeof(unsigned long));

    if (number == 0) {
        return data;
    } else if (number >= 0 && number <= 255) {
        data[0] = number;
        return data;
    } else {
        add_long(&data[0], number);
        return data;
    }
}

unsigned char *  AMQPWrapper::convertLong(long number)
{
    unsigned char * data = malloc(sizeof(long));

    if (number == 0) {
        return data;
    } else if (number >= -128 && number <= 127) {
        data[0] = number;
        return data;
    } else {
        add_long(&data[0], number);
        return data;
    }
}
